packages/testring-plugin-selenium-driver/src/plugin/index.ts

```diff
@@ -0,0 +1,484 @@
+import { IBrowserProxyPlugin } from '@testring/types';
+import { spawn } from '@testring/child-process';
+import { Config, Client, remote } from 'webdriverio';
+import { ChildProcess } from 'child_process';
+import { loggerClient } from '@testring/logger';
+
+
+const DEFAULT_CONFIG: Config = {
+    port: 4444,
+    desiredCapabilities: {
+        browserName: 'chrome'
+    }
+};
+
+function delay(timeout) {
+    return new Promise<void>((resolve) => setTimeout(() => resolve(), timeout));
+}
+
+export class SeleniumPlugin implements IBrowserProxyPlugin {
+
+    private browserClients: Map<string, Client<any>> = new Map();
+    private waitForReadyState: Promise<void> = Promise.resolve();
+    private localSelenium: ChildProcess;
+
+    constructor(private config: Config) {
+        if (this.config.host === undefined) {
+            this.runLocalSelenium();
+        }
+    }
+
+    private async createClient(applicant: string): Promise<any> {
+        if (this.browserClients.has(applicant)) {
+            return;
+        }
+
+        const client = remote({
+            ...DEFAULT_CONFIG,
+            ...this.config
+        } as any);
+
+        await client.init();
+
+        this.browserClients.set(applicant, client);
+    }
+
+    private getChromeDriverArgs() {
+        const chromeDriver = require('chromedriver');
+
+        return [`-Dwebdriver.chrome.driver=${chromeDriver.path}`];
+    }
+
+    async runLocalSelenium() {
+        this.waitForReadyState = delay(1000);
+        const seleniumServer = require('selenium-server');
+        const seleniumJarPath = seleniumServer.path;
+        let args = [
+            ...this.getChromeDriverArgs(),
+            '-jar', seleniumJarPath,
+            '-port', DEFAULT_CONFIG.port
+        ];
+
+        this.localSelenium = spawn('java', args);
+
+        this.localSelenium.stdout.on('data', (data) => {
+            loggerClient.debug(data.toString());
+        });
+
+        this.localSelenium.stderr.on('data', (data) => {
+            loggerClient.debug(data.toString());
+        });
+
+    }
+    public async end(applicant: string) {
```

Comment:
Gets kinda repetitive here

Reply:
I'd shorten it further and wrap it in async function which itself awaits for ready state and resloves client